Unify assembler structure

gassmoeller opened the pull request:
Our current assembler system is somewhat messy, because the structure for how an assembler has to look like is not well defined. Essentially any function can be attached to the available signals, and we do not know which functions are already attached (because we can not check which type each connection has), or which quantities an individual assembler needs to work correctly. This has lead to the extremely confusing logic inside of `set_assemblers`, and to the fact that for every new assembler we need to modify three objects: Connect to the associated signal. Add the parent object to `assembler_objects`. Modify the correct `assembler_properties` object to compute the correct quantities before calling the assembler.
If instead we had an interface-plugin system as for the other parts of ASPECT, all this would be handled in one object, the individual assembler plugin. It would have a defined function that is called by Simulator (which replaces the signal), and could (optionally) implement another function that specifies which quantities need to be provided (which replaces the properties object). We could still keep the `set_assemblers` signal that allows every part of the code to modify the list of existing assembler plugins with the added benefit that we can then remove specific plugins, instead of wiping them all and starting from scratch, which is our only option right now.
This is a very early draft on which changes we would need to implement in order to move in such a direction, but I wanted to start a discussion about it first, before spending more time with it. The important bits can be seen in `assembly.h` below. For now I moved the `StokesPreconditioner` assembly out of the `StokesAssembler` class and made it a separate fully-functional assembler plugin. It still connects its `execute` function to the old signal, but we could as well call it via the pointer to the base class stored in `assembler_objects`. The one disadvantage of the new system is that we have less flexibility of providing plugin-specific information into the assembler (see my workaround for `pressure_scaling` below). However, I would argue that we can provide all necessary information in two ways: Either add it to the `scratch` object that is already designed to store all necessary information or provide it via the `SimulatorAccess` class.

I would like to ask for opinion from @tjhei and @bangerth (and all other interested parties). I think the new way would keep the code cleaner and will be easier to understand. Everything else (performance, flexibility) should be similar to the current approach with signals.

bangerth commented:
Yes, I like this approach. I think it is fair to say that my idea with the signals has failed and collapsed under its own weight.

gassmoeller commented:
Do not be so harsh on your idea :smirk:. It works very well for simple functions. Our assemblers are just getting too complicated. And we should definitely keep the `set_assemblers` signal to allow users to modify the selected assemblers.
I will continue working on this after next week.

tjhei commented:
Yes, this is what I have been thinking about for a while. Should we delay this re-shuffling until after we are done with the whole Newton stuff?

gassmoeller commented:
Yes lets wait until after the newton assemblers are merged, I discussed this with @MFraters in #1784.

gassmoeller commented:
I am reasonably far with the rework of the assembler structure, but I would like to discuss a few interface questions before asking for a formal review, because they might change quite a few things:

1. The new `Assemblers::Interface` base class has one `execute` function that has parameters of type `ScratchBase *`. This allows every assembler to implement that single function, but it also means every assembler has to `dynamic_cast` the type of `scratch` to its appropriate scratch object (i.e. StokesPreconditioner, StokesSystem, or AdvectionSystem). I could instead declare three separate overloads for the execute function in `assemblers/interface`, with default implementations that throw assertions. This would shorten each Assembler, and it would also implicitly check that each assembler is used only in the assembly of the correct system. I would prefer the three function way, but wanted to ask for opinions before changing all those lines again (it would significantly shorten this PR, and also require less changes to user written assemblers).

2. The `scratch` and `data` parameters are conceptually similar to `MaterialModelInputs` and `MaterialModelOutputs`, however contrary to the latter, `scratch` is currently not handed over as `const`. This leads to the strange situation that the order of assembler calls can matter, if an assembler modifies the scratch object. Does anyone see a necessity for this behavior, or can I just make `scratch` `const`?

3. Many assemblers need a `cell`, and `face` (and some `face_no`) input, but they are currently not explicitly stored in `scratch`. They can be reconstructed from the `fe_(face_)values` object, but it is quite a bit of replicated code. Unless somebody objects I would prefer to store those explicitly in the scratch objects to make assemblers simpler.

gassmoeller commented:
I updated this PR to incorporate the Newton assemblers. This is ready for a review. 
Sorry for the long changelog, but it really mostly is a conversion of single functions to assembler objects.

bangerth commented:
There's now a merge conflict. Can you resolve it?

gassmoeller commented:
The merge conflict is resolved and the tests pass. I am not sure why the astyle tester fails in a file that this PR does not even touch though.

MFraters commented:
I had the same problem in #1953, that a unchanged file failed the astyle test. A rebased 'fixed' the problem. I am not sure what causes this problem.

gassmoeller commented:
Thanks Menno, rebasing did the trick. Can we move forward with this PR, and block other changes at the assemblers for the moment? Every time something changes in the assemblers there will be a merge conflict with this PR, and I feel this would really improve the structure of our assembler system.

bangerth commented:
This is difficult to review for me because I haven't paid attention to how the assemblers are built these days. The patch is already pretty unreadable :-) (Not because it's poorly written, but because it moves so much code around.)

Do any of the other reviewers have a better overview of how this all works and could do the review?

jdannberg commented:
I'll have a look tomorrow.

jdannberg commented on include/aspect/simulator/assemblers/advection.h:
```diff
@@ -0,0 +1,102 @@
+     * equation for the current cell.
+     */
+    template <int dim>
+    class AdvectionSystem : public Assemblers::Interface<dim>,
```
One point that confused me about this design, and I guess we can talk about that, is that now, if you call an assembler, you don't know any more if that assembler changes the Stokes or the Advection system. Before, with the signals, they were all in one class that either said advection or Stokes or other or whatever, but now they are just in different files. So if the assembler is just called `AdvectionSystem`, that's totally fine, but if it's called something weird like `PhaseBoundaryAssembler`, it's suddenly not clear anymore what it does. 
So I guess we can either say we don't care, or we may want to think about some naming convention, or maybe you can think of another way for how to make this clearer (I think, in particular as the set_assemblers function is quite long, it could make sense to have names that are easy to understand).

jdannberg replied:
As discussed, one could also use different namespaces for the Stokes/Advection/Other assemblers. Then, the assembler would only have to be called `IncompressibleTerms` instead of `StokesIncompressibleTerms`, and would be in the Stokes namespace.

gassmoeller replied:
I like the namespace idea (disclaimer: it was mine) :smile:. If nobody objects, I will regroup the assemblers accordingly.

jdannberg commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -31,35 +30,18 @@ namespace aspect
+      const double pressure_scaling = scratch.pressure_scaling;
 
       // First loop over all dofs and find those that are in the Stokes system
       // save the component (pressure and dim velocities) each belongs to.
```
I discussed this with @gassmoeller, and this is not something for this pull request, but something that relates to #733 and could be addressed later. To avoid duplicating so much code in the different assemblers, we could move at least the lines below where things in the scratch object are filled out of the assembler, and into the `local_assemble_...` functions. Many of those things are similar for the all Stokes assemblers, or all Stokes preconditioners, and while we would have to fill the things needed by all assemblers in the `local_assemble_...` functions (so potentially also some things that wouldn't be needed later on), it might still not be slower in the case of multiple assemblers being used together, as we would have to go through the loops only once. 
And it would have the advantage that the assemblers would be quite a bit shorter.

jdannberg commented on include/aspect/simulator/assemblers/stokes.h:
```diff
@@ -0,0 +1,204 @@
+     * This class assembles the right-hand-side term of the Stokes equation
+     * that is caused by the compressibility in the mass conservation equation.
+     * This class approximates this term as
+     * $ - \nabla \mathbf{u} = \frac{1}{\rho} * \frac{\partial rho}{\partial p} \rho \mathbf{g} \cdot \mathbf{u}$
```
That is not the correct documentation.

jdannberg commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,629 @@
+         * get_additional_output() returns an instance before adding a new
+         * one to the additional_outputs vector.
+         */
+        virtual void create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &) const {}
```
I don't really understand how this works now. Before, this was a function that was part of the simulator, and now you moved it to the assembler interface, but only some assemblers implement it? For example, for Stokes only the StokesIncompressibleTerms. What if I don't use that assembler?
And you also changed some of the test results that tested if this is working correctly, so it would be great if you could explain this a bit more.

gassmoeller replied:
If you do not use this assembler then this additional material model outputs object does not get created. But that is fine, each Assembler is only responsible for creating the additional material model outputs that it actually uses during assembly. This also means each assembler  has to check if the relevant additional material model outputs already exists (because another assembler already created them).

jdannberg commented on source/simulator/assembly.cc:
```diff
@@ -525,311 +102,157 @@ namespace aspect
-    namespace OtherTerms
+    template <int dim, class AssemblerType>
+    void
+    initialize_simulator(Simulator<dim> &simulator,
```
Can you document that function?

jdannberg commented on source/simulator/assembly.cc:
```diff
@@ -973,13 +367,17 @@ namespace aspect
+    initialize_simulator(*this,assemblers->advection_system);
+    initialize_simulator(*this,assemblers->advection_system_on_boundary_face);
+    initialize_simulator(*this,assemblers->advection_system_on_interior_face);
+    initialize_simulator(*this,assemblers->advection_system_residual);
```
Hm, this is not really nice. Isn't there a way to do this for all of them at once?

gassmoeller replied:
It turns out there is a way, but for that I need to implement a constructor for all assemblers (this constructor would take the `this` pointer to the simulator and call the constructor of the `SimulatorAccess` class that immediately does what `initialize_simulator` does). But this needs more lines than I have here. Still if you would prefer the constructor I can implement that (it might look cleaner, but require more lines).

jdannberg replied:
Hm. What kind of error message do you get if you forget to call this line for your assembler? I am just thinking, if you make a new type of assembler and forget to implement this line, how easy is it to find out what the problem is? If that is obvious, I would be okay with leaving it like it is.

jdannberg commented on source/simulator/assembly.cc:
```diff
@@ -1001,6 +399,9 @@ namespace aspect
     scratch.finite_element_values.reinit (cell);
+    scratch.cell = cell;
+    scratch.pressure_scaling = material_model->reference_viscosity()
+                               / geometry_model->length_scale();
```
As I mentioned earlier, it would be much better to use the simulator access for this instead.

jdannberg commented on source/simulator/assembly.cc:
```diff
@@ -1189,6 +592,9 @@ namespace aspect
     scratch.finite_element_values.reinit (cell);
+    scratch.cell = cell;
+    scratch.pressure_scaling = material_model->reference_viscosity()
+                               / geometry_model->length_scale();
```
Same point about the pressure scaling.

gassmoeller commented on include/aspect/simulator.h:
```diff
@@ -740,51 +740,13 @@ namespace aspect
-       * internal::Assemblers::AssemblerBase::create_additional_material_model_outputs()
-       * functions from each object in Simulator::assembler_objects.
-       */
-      void create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &) const;
```
I moved it over. Copy-Paste error.

gassmoeller commented:
I think I addressed all comments, except the discussion points I answered above.

tjhei commented on doc/update_source_files_to_2.0.0.sed:
```diff
@@ -62,3 +62,11 @@ s/tracer particle/particle/g
+s/struct AssemblerLists/class Manager/g
+s/AssemblerLists/Manager/g
+s/AssemblerBase/Interface/g
+s:assembly.h:simulator/assemblers/interface.h:g
```
we could keep assembly.h (that includes the other files) to avoid this change?!

gassmoeller replied:
I would vote for removing the file. The interface is broken anyway for anybody who wrote an individual assembler (not many people), and those would be the only ones including assembly.h. interface.h *is* the new assembly.h.

tjhei commented on include/aspect/free_surface.h:
```diff
@@ -23,12 +23,30 @@
+    class ApplyStabilization: public Assemblers::Interface<dim>,
+      public SimulatorAccess<dim>
+    {
+        void execute (internal::Assembly::Scratch::ScratchBase<dim>   &scratch,
```
why does this work without "public:" ?

tjhei commented on include/aspect/melt.h:
```diff
@@ -137,132 +137,132 @@ namespace aspect
       public SimulatorAccess<dim>
     {
       public:
+        virtual ~MeltInterface () {};
```
I don't think you need the destructor here as this is still an abstract base class

tjhei commented on include/aspect/newton.h:
```diff
@@ -113,88 +113,129 @@ namespace aspect
+      public SimulatorAccess<dim>
+    {
+      public:
+        virtual ~NewtonStokesReferenceDensityCompressibilityTerm () {};
```
remove semicolon. Do we normally have all those in the .cc files? I don't mind having them here...

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+  {
+    namespace Assembly
+    {
+      namespace Scratch
```
My vote would be to remove this namespace as it doesn't convey additional information (``Scratch::ScratchBase`` vs ``ScratchBase``)

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+            :
+            cell(),
+            face_number(numbers::invalid_unsigned_int)
+          {};
```
redundant semicolons here and below.

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+        template <int dim>
+        struct CopyDataBase
+        {
+          CopyDataBase() {};
```
You don't need this constructor.

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+        {
+          CopyDataBase() {};
+
+          virtual ~CopyDataBase () {};
```
redundant semicolon. Shouldn't this destructor be pure virtual (=0)?

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+    class Interface
+    {
+      public:
+        virtual ~Interface () {}
```
make pure virtual?

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+        virtual
+        std::vector<double>
+        compute_residual(internal::Assembly::Scratch::ScratchBase<dim> &) const
+        {
```
move implementation to .cc?

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+         * equation (i.e. which terms are actually included in the
+         * equation). Thus different objects compute different residuals (i.e.
+         * the residual for a melt advection equation looks different from the
+         * residual for a passive compositional field).
```
add "For assemblers for the Stokes system, an implementation of this function is not necessary"

tjhei commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -0,0 +1,718 @@
+        {
+          AssertThrow(false, ExcMessage("This function should either be implemented "
+                                        "(if this is an assembler that has to compute a residual, or not be called "
+                                        "(if this is an assembler that has not to compute a residual)."));
```
copy&paste error?

tjhei commented on include/aspect/simulator_access.h:
```diff
@@ -645,6 +639,13 @@ namespace aspect
       get_newton_handler () const;
 
+      /**
+       * Return a set of boundary indicators that describes which of the
```
copy&paste comment?

tjhei commented on include/aspect/simulator_signals.h:
```diff
@@ -263,8 +257,7 @@ namespace aspect
     boost::signals2::signal<void (const SimulatorAccess<dim> &,
-                                  aspect::internal::Assembly::AssemblerLists<dim> &,
-                                  std::vector<std_cxx11::shared_ptr<internal::Assembly::Assemblers::AssemblerBase<dim> > > &)>
+                                  aspect::Assemblers::Manager<dim> &)>
```
I guess we can not avoid this breaking change. Do we care?

gassmoeller replied:
No. We are in for a major release. Moreover: Anybody who used this interface used it to connect a function to a signal that now no longer exists, so they have to change their implementation anyway, we might as well fix the interface.

tjhei commented on source/simulator/assemblers/advection.cc:
```diff
@@ -437,11 +446,6 @@ namespace aspect
                 }
             }
         }
-      else if (cell->has_periodic_neighbor (face_no))
```
why remove this?

gassmoeller replied:
This logic is now in assembly.cc:1040 (i.e. before we call the assembler for a face we check if there is a periodic neighbor, and then immediately call the other assembler).

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -525,311 +102,160 @@ namespace aspect
+    // lines in set_assemblers(), where this operation appears multiple times.
+    template <int dim, class AssemblerType>
+    void
+    initialize_simulator(Simulator<dim> &simulator,
```
can you make the simulator const here?

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -525,311 +102,160 @@ namespace aspect
-     */
-    namespace OtherTerms
+    // This function initializes the simulator access for all assemblers
+    // inside of the assemblers parameter. This is just a shortcut to save some
```
very cool!

tjhei commented on source/simulator/entropy_viscosity.cc:
```diff
@@ -128,9 +128,15 @@ namespace aspect
+    for (unsigned int i=1; i<assemblers->advection_system.size(); ++i)
+      {
+        std::vector<double> new_residual = assemblers->advection_system[i]->compute_residual(scratch);
+        for (unsigned int i=0; i<residual.size(); ++i)
```
can you please use a different parameter than reusing ``i``?

tjhei commented on tests/additional_outputs_02/screen-output:
```diff
@@ -8,24 +8,14 @@ Number of active cells: 4 (on 2 levels)
 Number of degrees of freedom: 84 (50+9+25)
 
 *** Timestep 0:  t=0 seconds
-* create_additional_material_model_outputs() called
```
why is this no longer called?

gassmoeller replied:
Because previously if an assembler was in the list of assembler_objects create_additional_material_model_outputs() would be called for all equations, while now it is only called for the equations in which this assembler is actually used (in this case the Stokes equations). Since the Assembler can only be used in the Stokes equation (to successfully cast the scratch object), I can not reproduce the old output, but the new one is more reasonable anyway.

tjhei commented on tests/additional_outputs_03/screen-output:
```diff
@@ -8,32 +8,14 @@ Number of active cells: 4 (on 2 levels)
 Number of degrees of freedom: 84 (50+9+25)
 
 *** Timestep 0:  t=0 seconds
-* create_additional_material_model_outputs() called
```
same here?

gassmoeller commented:
I think I addressed most comments. About the destructor implementations: I agree that we usually move those into the .cc file. It feels like a serious case of wasted lines of code in this case though. Each of the ~20 assemblers only implements one single function that is `const`, and none has any member variables, so having a destructor is kind of useless anyway. I would vote for keeping it in the .h files.
